### ISSUE TITLE
[feature request] ignore-old-news

### ISSUE THREAD

**Issue by emanruse** [newsboat/newsboat#1419]
Quite often I face the following situation:

1. Starting point (example article list of a feed):
```
N 2018-01-01 Article1
N 2018-02-05 Article2
N 2019-09-08 Article3
...
N 2020-01-01 ArticleN-3
N 2020-01-02 ArticleN-2
N 2020-01-03 ArticleN-1
N 2020-01-05 ArticleN
```

2. I read all articles, delete the read ones (perhaps keep a few read ones). Result:
```
  2020-01-01 ArticleN-3
  2020-01-02 ArticleN-2
```

3. Some time passes (days, weeks, not a determinate period) and for some reason the server feed updates its file. Fetching that file during feed update "restores" all the N articles which I deleted earlier and shows them as unread. Only the ones I kept as read remain as I kept them and if new ones were added server side, they show at the end of the article list too:

```
N 2018-01-01 Article1
N 2018-02-05 Article2
N 2019-09-08 Article3
...
  2020-01-01 ArticleN-3
  2020-01-02 ArticleN-2
N 2020-01-03 ArticleN-1
N 2020-01-05 ArticleN
N 2020-01-06 ArticleN+1
N 2020-01-07 ArticleN+2
...
```

This is really really annoying. I understand that this is maybe not newsboat specific, as I have seen it in other readerers too. The workaround is:

- (overkill) keep all the hundreds of read articles forever
- always keep the latest article as unread. Then if that happens, I delete the older ones (taking care not to delete the ones you want to keep, still confusing and takes time).

This doesn't happen for all feeds, but for some it does happen from time to time (at different times). It would be great to have a feature like (config line):

```
ignore-old-news "yes"     # I believe this should be the default value
```

This would prevent old deleted articles to pop as new unread ones. The algorithm would be (pseudo code):
```
while (newly_fetched_article) {
	if (article does not exist locally) {
		// feed_last_update_time must be stored
		// in local db/cache for each feed
		if (article_date < feed_last_update_time) {
			if (ignore_old_news == "yes") {
				continue;
			}
		}
		store_article_as_unread();
	}
}
feed_last_update_time=current_time;
save_in_local_cache(feed_last_update_time);
// continue to next feed
```

**Comment by Minoru** [comment]
Thanks for opening the issue! Before discussing any fixes, though, let's figure out *why* this even happens.

Can you share your config? `max-items` and `keep-articles-days` may well lead to this effect.

Can you share URLs of the feeds where you observed this most often, and most recently? A perfect feed would be one that will let me reproduce this as soon as possible.

**Comment by emanruse** [comment]
> Before discussing any fixes, though, let's figure out *why* this even happens.

As I tried to explain, it is my speculation that the server side feed file changes in a way which tells the reader "hey, I am a new article (though my date is from 3 years ago)". I have not looked into the details of the feed file, but I speculate there may be an attribute saying "this article was updated on date X" (where X != actual article content date).

> Can you share your config?

```
cleanup-on-quit yes
scrolloff 10
article-sort-order date asc
feed-sort-order firsttag
bind-key j down
bind-key k up
bind-key S save-all
datetime-format "%d"
goto-next-feed no
save-path "~/saved-articles/"
articlelist-format "%f %D %?T?|%-17T| ?%t"
articlelist-title-format "%T (%u unread, %t total)"
itemview-title-format "'%T' (%u unread, %t total)"
feedlist-title-format "(%u unread, %t total)%?T? - tag ‘%T’&?"
show-read-feeds no
ignore-mode "download"
```

> `max-items` and `keep-articles-days` may well lead to this effect.

But these have a completely different purpose and may actually have undesired effects.

> Can you share URLs of the feeds where you observed this most often, and most recently? A perfect feed would be one that will let me reproduce this as soon as possible.

I don't keep an accurate record of how often this happens with each feed. I have noticed though that it happens quite often with some of the feeds generated by the Facebook Bridge | Main Site and Twitter Bridge on https://rss.nixnet.services/ (through which I follow some public Facebook and Twitter pages). You can give it a try with FB/TW pages of your choice. For me it often displays deleted articles from previous dates.

**Comment by Minoru** [comment]
> As I tried to explain, it is my speculation that the server side feed file changes in a way which tells the reader "hey, I am a new article (though my date is from 3 years ago)".

This is correct: RSS items contain a `<guid>` tag, and Atom has `<id>`; these are used to tell the items apart. Newsboat considers an article new if it doesn't have its guid/id in the database — no matter the dates, content, or anything else.

> I have not looked into the details of the feed file, but I speculate there may be an attribute saying "this article was updated on date X" (where X != actual article content date).

There are `<published>` and `<updated>` tags in Atom, and similar one(s) in RSS (don't remember exactly), but I don't think they could cause this — Newsboat simply copies the dates to its database.

Anyway, I set up a shell loop that fetches https://rss.nixnet.services/?action=display&bridge=Twitter&context=By+username&u=elonmusk&format=Atom every 5 minutes and records the file in a Git repository. I also added that feed to my Newsboat, and will delete all the articles that appear. Hopefully this would be enough to reproduce this, and then I'll use the Git repo to find out exactly what changed in the feed.

**Comment by Minoru** [comment]
Oh, forgot to ask one more question: your config contains `ignore-mode "download"`, but no `ignore-article` rules. Is this the entire config? I'm asking because changing the rules might well lead to this behaviour if they admit some articles that the previous rules ignored.

**Comment by emanruse** [comment]
> Oh, forgot to ask one more question: your config contains `ignore-mode "download"`, but no `ignore-article` rules. Is this the entire config? I'm asking because changing the rules might well lead to this behaviour if they admit some articles that the previous rules ignored.

I have been experiencing this long before learning about the ignore feature.

And no, that is not the entire config. I can't share everything for privacy reasons. However ignoring is unrelated and I am confident my ignore rules have nothing to do with this. Also they are for other feeds (not the ones in which I have noticed the issue).

**Comment by Minoru** [comment]
> And no, that is not the entire config.

I wish you had the courtesy to point that out when you posted the config. The closer my setup is to yours, the higher the chance of me reproducing the bug, and thus the chance of fixing it.

Since you can't share your exact setup, you'd have to work a bit to create a setup which you *can* share. Start with an empty config, add a feed you don't personally care about, describe that setup *in detail* — exact feed URL, complete config, anything else you *intentionally* configured in relation to Newsboat — so I can reproduce it, and we'll continue from there. You can run this in parallel with your other Newsboat using `--config-file`, `--cache-file`, and `--url-file` command-line parameters.

**Comment by emanruse** [comment]
As I explained, this has nothing to do with ignore rules or specific config. I have been using newsboat for about a year and this has been an issue from the very beginning. It was an issue even with the most basic empty config. So the feature request related to it has absolutely nothing to do with a particular config.

Please also don't forget what I wrote earlier: I don't think this is an issue with a particular reader. As [it seems] we agreed: there are XML attributes which may be the reason for this. So there is no need to look for the reason where it surely is not (in config).

FWIW recently Purism's feed "updated" all it's old news since 2014 too (https://puri.sm/feed/). This seems to happen with some WordPress generated feeds from time to time. I have seen it happening with GitHub's release feeds too (e.g. https://github.com/gorhill/uBlock/releases.atom). Again: I don't keep an exact record of when it happens with a particular feed. Just sharing what comes to my mind as I am typing this. I have no idea what causes this server-side. It might be some cache or db update of the CMS which generates the feed - I don't know. The fact is it happens and it is not newsboat specific. But it can be worked around with the suggested feature. Hence the request.

I hope that clarifies.

**Comment by Minoru** [comment]
I'm sorry this thread got so antagonistic, but it feels like you've made up your mind about what should be done, and my inquiries sound like a waste of time to you. However, I'd prefer to fix problems instead of working around them, so I'd like to find what causes this instead of papering over it. It might turn out that the proposed feature is indeed the best solution, but it's impossible to tell without knowing what the root cause is.

Purism feed is a valuable detail because it's archived on the Wayback Machine. Do you happen to remember more precisely when the update happened? The feed is archived [about every other week](https://web.archive.org/web/2020*/https://puri.sm/feed/), so remembering a month when it happened would be good enough.

**Comment by Minoru** [comment]
I think I've found the root cause: the GUIDs of the items are derived from the links to their web versions, and those links changed from http:// to https://:

- 2015.02.02 — GUIDs use http://, including the items from 2014
- 2016.11.26 — items from 2014 are removed from the feed
- 2017.01.05 — items from 2014 still absent
- 2018.01.25 — ditto
- 2019.01.04 — items from 2014 appear again, but their GUIDs now use HTTPS

I'll re-visit the proposed feature once I got over the negativity from interactions here.

**Comment by emanruse** [comment]
Great to know you found the reason! (or at least one of them because some feeds don't have guid attributes, do they?)

Forgive me if anything I said sounded uncooperative. That was not my intention at all.

As for Purism and when I noticed it last - IIRC it was either 2021-January or 2020-December, but I am not 100% sure.
